File: range_sum_of_BST.py

# https://leetcode.com/problems/range-sum-of-bst/

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
	# O(n) time, O(n) space
	def rangeSumBST(self, root: TreeNode, L: int, R: int) -> int:
		def aux(root):
			if not root: return 0
			if root.val < L:
				return aux(root.right)
			if R < root.val:
				return  aux(root.left)
			return root.val + aux(root.left) + aux(root.right)

		return aux(root)

File: range_sum_of_bst.cpp
#include <iostream>
#include "TreeNode_util.h"

// https://leetcode.com/problems/range-sum-of-bst/

class Solution {
public:
	// O(n) time.
	// O(h) space (recursion stack)
	// where n = number of nodes in the tree, 
	// h = height of the tree.
	int rangeSumBST(TreeNode* root, int low, int high) {
		if(not root)
			return 0;

		if(high < root->val)
			return rangeSumBST(root->left, low, high);

		if(root->val < low)
			return rangeSumBST(root->right, low, high);

		return root->val + rangeSumBST(root->left, low, high) 
						 + rangeSumBST(root->right, low, high);
	}
};
